### PR TITLE
Tweak image strategy

### PR DESCRIPTION
**Pull request by variadico**
Based on feedback from docker-library/official-images, we're reverting
the use of multi-stage builds because they don't work as you'd expect
in docker-library/official-images.

In this change, we'll use the CGO enabled official release binaries for
all containers that support CGO. For images that don't support CGO, such
as scratch, we'll use special locally built binaries in this repo with
CGO disabled, like before.

While, we do partially revert to the original behavior. This still helps long
term because it lays the framework for adding new images. Scratch is the only
odd ball that will need binaries built. All future Linux images will only need
to download binaries.

### CODE REVIEW

**Comment by variadico** [comment]
I'm testing on the Docker side with this PR.
https://github.com/docker-library/official-images/pull/6766

CI is green on that side.

**Comment by derekcollison** [comment]
Will let Ivan weigh in tomorrow when he is back.

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -3,33 +3,31 @@ on: push
...
       run: |
-        cd 2.1.0/scratch
-        docker build --tag nats:2.1.0-scratch .
-        run_id=$(docker run --detach nats:2.1.0-scratch)
-        sleep 1
-        test -n "$(docker ps --filter "id=${run_id}" --filter "status=running" --quiet)"
-        docker kill "$run_id"
-  windows:
-    runs-on: windows-latest
+        ./scripts/build-images.sh
```
Comment (kozlovic) [inline .github/workflows/main.yaml]: From what I understand this is building the image just for test. Could we call this `./scripts/test-build-images.sh`.

File: .github/workflows/main.yaml [hunk 2]
```diff
@@ -3,33 +3,31 @@ on: push
...
+    - name: Test images
+      run: |
+        ./scripts/test-images.sh
```
Comment (kozlovic) [inline .github/workflows/main.yaml]: Similarly: `./scripts/test-run-images.sh`?

File: .github/workflows/main.yaml [hunk 3]
```diff
@@ -3,33 +3,31 @@ on: push
...
-        cd 2.1.0/nanoserver1803
-        docker build --tag nats:2.1.0-nanoserver1803 .
-    - name: Test nanoserver1809
+    - name: Build images
+      shell: powershell
       run: |
-        cd 2.1.0/nanoserver1809
-        docker build --tag nats:2.1.0-nanoserver1809 .
-    - name: Test servercoreltsc2016
+        ./scripts/build-images-2019.ps1
```
Comment (kozlovic) [inline .github/workflows/main.yaml]: test-build-images-2019.ps1?

File: .github/workflows/main.yaml [hunk 4]
```diff
@@ -3,33 +3,31 @@ on: push
...
+    runs-on: windows-2016
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v1
+    - name: Build images
+      shell: powershell
       run: |
-        cd 2.1.0/servercoreltsc2016
-        docker build --tag nats:2.1.0-servercoreltsc2016 .
+        ./scripts/build-images-2016.ps1
```
Comment (kozlovic) [inline .github/workflows/main.yaml]: test-build-images..?

File: 2.1.0/scratch/get-scratch-bins.sh [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	# format that the Dockerfiles have for specifying the version.
+	ver=( NATS_SERVER 2.1.0 )
+	git clone --branch "v${ver[1]}" --depth 1 https://github.com/nats-io/nats-server.git .
+
+	export CGO_ENABLED=0
+	export GO111MODULE=off
+
+	commit="$(git rev-parse --short HEAD)"
+
+	GOARCH=amd64 go build -o /out/amd64/nats-server -v -a -tags netgo -installsuffix netgo \
```
Comment (kozlovic) [inline 2.1.0/scratch/get-scratch-bins.sh]: Images are built and put into `/out/..` but that's not where the server exes are located. They are in 2.1.0/scratch/<arch>?

File: 2.1.0/nanoserver1803/Dockerfile [hunk 6]
```diff
@@ -1,20 +1,10 @@
...
-	go build -o C:\nats-server.exe -v -a -tags netgo -installsuffix netgo -ldflags "$ldflags";
-
-# Final stage
 FROM mcr.microsoft.com/windows/nanoserver:1803
-ENV NATS_DOCKERIZED=1
-COPY --from=tmp C:\\nats-server.exe C:\\nats-server.exe
+
+ENV NATS_DOCKERIZED 1
+
+COPY --from=nats:2.1.0-servercore1803 C:\\nats-server.exe C:\\nats-server.exe
```
Comment (kozlovic) [inline 2.1.0/nanoserver1803/Dockerfile]: That confused me: So it means that the servercore1803 needs to be built first, then the nanoserver1803 is getting the executable from servercore1803 image and adding it here. Hope that works :-)

File: scripts/build-images-2019.ps1 [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+# The servercore images must be built before the nanoserver images.
+cd 2.1.0/servercore1809
+Write-Host "building servercore1809"
+docker build --tag nats:2.1.0-servercore1809 .
+cd ../nanoserver1809
+Write-Host "building nanoserver1809"
+docker build --tag nats:2.1.0-nanoserver1809 .
+
+# We can't test these images on GitHub CI. We get the following error on both
```
Comment (kozlovic) [inline scripts/build-images-2019.ps1]: Still true? You have made some changes in the official images PR by adding a constraint on servercore.

Comment (variadico) [reply scripts/build-images-2019.ps1]: Yeah, I think so. The constraint stuff is only for Docker's CI. I think the problem is that our GitHub Actions CI doesn't have the right version of Windows available. We only get windows-2016 or windows-2019, but I think we need something like windows-1803.

Not entirely sure, but it seems like it based on this.
https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility

File: 2.1.0/nanoserver1803/Dockerfile (hunk 6 above)
Comment (variadico) [reply 2.1.0/nanoserver1803/Dockerfile]: Yeah, this also confused me. But, it's how the Docker guys suggested to do it.
https://github.com/docker-library/official-images/pulls

¯\_(ツ)_/¯

File: 2.1.0/scratch/get-scratch-bins.sh (hunk 5 above)
Comment (variadico) [reply 2.1.0/scratch/get-scratch-bins.sh]: `get-scratch-bins.sh` is meant to be run in the `scratch` directory. The Go container is started with

```
--volume $(pwd):/out
```

which means that the current directory, which should contain the different architecture folders, will be mounted as `/out` in the container.

Then, in the container binaries get written to `/out`. So the mapping is like this.

```
host       container
./amd64 -> /out/amd64
./arm64v8 -> /out/arm64v8
```

I added a check to the script to make sure it doesn't get run from the top level of the repo.

Comment (kozlovic) [reply 2.1.0/scratch/get-scratch-bins.sh]: I missed that. Thanks!

**Pull request merged**